src/cookbooks/clients_last_seen_bits.md

```diff
@@ -0,0 +1,306 @@
+  BIT_COUNT(udf.bits28_range(days_seen_bits, 13, 0, 7)) > 0 AS active_in_week_0
+  BIT_COUNT(udf.bits28_range(days_seen_bits, 13, 7, 7)) > 0 AS active_in_week_1
+FROM
+  telemetry.clients_last_seen
+```
+
+The `bits28_range` function is powerful and flexible, but a bit cumbersome.
+We provide a `bits28_retention` convenience function that returns a nested
+structure with the standard retention definitions precalculated:
+
+```
+SELECT
+  submission_date,
+  udf.bits28_retention(days_seen_bits, submission_date).day_13.*
+FROM
+  telemetry.clients_last_seen
+
+/*
+| submission_date | metric_date | active_on_day_0 | active_in_week_0 | active_in_week_1 |
+|      2020-01-28 |  2020-01-15 | true            | true             | true             |
+*/
+```
+
+The `bits28_retention` struct also has `day_21` and `day_28` fields that can
+be used to calculate "2-Week Retention" and "3-Week Retention".
+
+### Full example query for 1-Week Retention variants
+
+```
+WITH base AS (
+  SELECT
+    *,
+    udf.bits28_retention(days_seen_bits, submission_date) AS retention,
+    udf.bits28_days_since_seen(days_created_profile_bits) = 13 AS is_new_profile
+  FROM
+    telemetry.clients_last_seen )
+SELECT
+  retention.day_13.metric_date,
+
+  -- 1-Week Retention matching GUD.
+  SAFE_DIVIDE(
+    COUNTIF(retention.day_13.active_in_week_1),
+    COUNTIF(retention.day_13.active_in_week_0)
+  ) AS retention_1_week,
+
+  -- 1-Week New Profile Retention matching GUD.
+  SAFE_DIVIDE(
+    COUNTIF(is_new_profile AND retention.day_13.active_in_week_1),
+    COUNTIF(is_new_profile)
+  ) AS retention_1_week_new_profile,
+
+  -- A more restrictive 1-Week Retention definition that considers only clients
+  -- active on day 0 rather than clients active on any day in week 0.
+  SAFE_DIVIDE(
+    COUNTIF(retention.day_13.active_in_week_1),
+    COUNTIF(retention.day_13.active_on_metric_date)
+  ) AS retention_1_week_active_on_day_0,
+
+FROM
+  base
+GROUP BY
+  metric_date
+WHERE
+  submission_date = '2020-01-28'
+```
+
+### N-day Retention
+
+Let's define `n`-day retention as the fraction of clients active on a given day
+who are also active within the next `n` days. For example, 3-day retention would
+have a denominator of all clients active on day 0 and a numerator of all clients
+active on day 0 who were also active on days 1 or 2.
+
+To calculate `n`-day retention, we need to use the lower-level `bits28_range`
+function:
+
+```
+DECLARE n INT64;
+SET n = 3;
+
+WITH base AS (
+  SELECT
+    *,
+    udf.bits28_range(days_seen_bits, n, 0, 1) AS seen_on_day_0,
+    udf.bits28_range(days_seen_bits, n, 1, n - 1) AS seen_after_day_0
+  FROM
+    telemetry.clients_last_seen )
+SELECT
+  DATE_SUB(submission_date, INTERVAL n DAYS) AS metric_date,
+  SAFE_DIVIDE(
+    COUNTIF(seen_on_day_0 AND seen_after_day_0),
+    COUNTIF(seen_on_day_0)
+  ) AS retention_n_day
+FROM
+  base
+GROUP BY
+  metric_date
+WHERE
+  submission_date = '2020-01-28'
+```
+
+### Retention using activity date
```

Comment:
This is getting fairly speculative here. I've never actually run any queries using these activity date-based bit patterns, so they're unvetted. But I'm fairly confident that the kind of calculation described here is feasible.

Reply:
Do we want to keep this or do we want to fairly firmly be in the submission_date camp?

Reply:
I want to have some advanced examples. I've added more of a disclaimer here about how it's unofficial.